Add tracks dropdown implementation

dippindots opened the pull request:
Fix cBioPortal/cbioportal#7934

Describe changes proposed in this pull request:
- Merge clinical tracks and heatmap tracks to `add tracks` dropdown

Previous:
![image](https://user-images.githubusercontent.com/15748980/106625760-fa82f500-6544-11eb-9f88-a5bb0e6758a1.png)

Current:
![image](https://user-images.githubusercontent.com/15748980/107722872-f32eba80-6cad-11eb-89d0-b6c5f300ab41.png)

Testing:
[TEST LINK](https://deploy-preview-3622--cbioportalfrontend.netlify.app/results/oncoprint?genetic_profile_ids_PROFILE_MUTATION_EXTENDED=ccle_broad_2019_mutations&genetic_profile_ids_PROFILE_COPY_NUMBER_ALTERATION=ccle_broad_2019_cna&cancer_study_list=ccle_broad_2019&Z_SCORE_THRESHOLD=2.0&RPPA_SCORE_THRESHOLD=2.0&data_priority=0&profileFilter=0&case_set_id=ccle_broad_2019_cnaseq&gene_list=TP53&geneset_list=%20&tab_index=tab_visualize&Action=Submit)

inodb commented:
@tmazor @schultzn could you take a look? We plan to merge this by Monday - thanks!

dippindots commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -5063,4 +5061,112 @@ export class ResultsViewPageStore {
         }
     }
+
+    readonly clinicalTrackOptions = remoteData({
```
moved from `AddClinicalTracks.tsx`

adamabeshouse replied:
why not put these into `AddTracks.tsx`?

dippindots replied:
@adamabeshouse good point, I will move them to `AddTracks.tsx`

dippindots commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -5063,4 +5061,112 @@ export class ResultsViewPageStore {
+        },
+    });
+
+    readonly clinicalAttributeIdToAvailableFrequency = remoteData({
```
moved from `AddClinicalTracks.tsx`

adamabeshouse replied:
why not into `AddTracks.tsx`?

dippindots commented on src/pages/resultsView/oncoprint/AddTracks.tsx:
```diff
@@ -0,0 +1,297 @@
+                  deriveDisplayTextFromGenericAssayType(tab.key as string)
+              ).join().length
+            : 0;
+        const genericAssayTabsWidth =
```
It's difficult to get reference of the whole tabs width, use the estimate width calculated by count of characters.

adamabeshouse replied:
we have a method that measures this, see `getTextWidth` in `packages/cbioportal-frontend-commons/src/lib/TextTruncationUtils.ts`

alisman replied:
seems fine to me

alisman replied:
listen to @adamabeshouse

dippindots replied:
sure, will do

dippindots commented on src/shared/components/oncoprint/ResultsViewOncoprint.tsx:
```diff
@@ -503,15 +506,13 @@ export default class ResultsViewOncoprint extends React.Component<
             get selectedHeatmapProfileId() {
                 return self.selectedHeatmapProfileId;
             },
-            get selectedHeatmapProfileAlterationType() {
-                return self.selectedHeatmapProfileAlterationType;
-            },
-            get selectedHeatmapProfileGenericAssayType() {
-                return self.selectedHeatmapProfileGenericAssayType;
-            },
```
It's safe to remove these, only used by Generic Assay previously.

dippindots commented:
@alisman @adamabeshouse, could I get your review on this pr? @adamabeshouse There is a `TODO` comment in this pr, we should add another method there to add category tracks or heatmap tracks based on data type.

tmazor commented:
Nice! A few thoughts:

1) The generic assay options should include a number like the other ones do:
![image](https://user-images.githubusercontent.com/10508276/107086434-016d6a00-67c8-11eb-9bff-1943f8d546fb.png)

2) I think we can do better than 'Heatmap' as the second tab name. It doesn't clearly explain what types of data can be found there, and is potentially confusing because treatment/microbiome/etc all load as heatmaps also. I think this tab will include mRNA/protein expression and DNA methylation, right? So we could call it 'Molecular' or 'mRNA/Protein/DNA Methylation' or 'Expression/DNA Methylation' or... hopefully someone else has a better idea? @schultzn @jjgao any thoughts?

Besides that, looks good!

alisman commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -3751,15 +3751,14 @@ export class ResultsViewPageStore {
                 _.filter(this.molecularProfilesInStudies.result!, profile => {
                     return (
                         ((profile.molecularAlterationType === MRNA_EXPRESSION ||
```
this whole condition could use a comment to explain why

alisman commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -4010,14 +4005,17 @@ export class ResultsViewPageStore {
 
     readonly genericAssayEntitiesGroupByGenericAssayType = remoteData<{
```
GroupedBy?

alisman commented on src/pages/resultsView/ResultsViewPageStore.ts:
```diff
@@ -5063,4 +5061,112 @@ export class ResultsViewPageStore {
+            this.clinicalAttributes_customCharts,
+        ],
+        invoke: () => {
+            const uniqueAttributes = _.uniqBy(
```
we should really move this logic to helper files.  putting all this inside the invoke is one reason by this file is >5000 lines long.   also it can't be tested

jjgao commented:
> I think we can do better than 'Heatmap' as the second tab name. It doesn't clearly explain what types of data can be found there, and is potentially confusing because treatment/microbiome/etc all load as heatmaps also. I think this tab will include mRNA/protein expression and DNA methylation, right? So we could call it 'Molecular' or 'mRNA/Protein/DNA Methylation' or 'Expression/DNA Methylation' or... hopefully someone else has a better idea? @schultzn @jjgao any thoughts?

@tmazor great point. Maybe we should list the tab based on profile type? ... if we can manage to squeeze many types in the UI.

dippindots commented:
@tmazor @jjgao Maybe we can use the same concept (`Gene Specific`) like what we are using in `StudyView`?
![image](https://user-images.githubusercontent.com/15748980/107094426-f967f700-67d4-11eb-87db-291d85b8dce3.png)

alisman commented on src/pages/studyView/addChartButton/genericAssaySelection/GenericAssaySelection.tsx:
```diff
@@ -29,35 +38,59 @@ export default class GenericAssaySelection extends React.Component<
+            } else {
+                const option = this.selectedProfileOption as ISelectOption;
+                // select profile if onSelectGenericAssayProfile exists
+                this.props.onSelectGenericAssayProfile &&
```
it's kind of weird that we pass in 2 handlers (onSelectGenericAssayProfile and onSubmit).    Is there a reason?

alisman commented on src/pages/studyView/addChartButton/genericAssaySelection/GenericAssaySelection.tsx:
```diff
@@ -85,20 +118,30 @@ export default class GenericAssaySelection extends React.Component<
     @computed
-    private get entityNotSelected() {
-        return _.isEmpty(this._selectedGenericAssayEntityIds);
+    private get buttonDisabled() {
```
maybe put a comment explaining why we only disable button in studyview and not results view

alisman replied:
this is the kind of knowledge that another developer cannot figure out

alisman commented on src/pages/studyView/addChartButton/genericAssaySelection/GenericAssaySelection.tsx:
```diff
@@ -85,20 +118,30 @@ export default class GenericAssaySelection extends React.Component<
-                    profileName: option.label,
-                };
-            });
+            if (this.props.studyViewMode) {
```
i don't love this studyViewMode thing.  Especially b/c this componont lives in studyView folder.  Is there a way we could keep this fork outside of the component?   note that it seems to consume ONLY from props, so why not just do it in parent or store?

alisman commented on src/shared/components/oncoprint/ResultsViewOncoprint.tsx:
```diff
@@ -224,6 +225,8 @@ export default class ResultsViewOncoprint extends React.Component<
     @observable selectedHeatmapProfileId = '';
     @observable heatmapGeneInputValue = '';
 
+    @observable selectedGenericAssayProfileId = '';
```
i see we do this above, but it's probably not a good idea to use empty string as essentially a null value.   why not just null?

alisman commented on src/shared/components/oncoprint/controls/CustomDropdown.tsx:
```diff
@@ -23,12 +24,12 @@ class CustomButton extends React.Component<any, {}> {
             <div
                 className={classNames('dropdown-menu', className)}
-                style={{ padding: '6px' }}
+                style={_.merge(styles, { padding: '6px' })}
```
i think you can do { ...styles, padding:'6px' }

alisman commented on src/shared/components/oncoprint/controls/CustomDropdown.tsx:
```diff
@@ -69,7 +70,11 @@ export default class CustomDropdown extends React.Component<
-                    <CustomMenu bsRole="menu" className={classNames(className)}>
+                    <CustomMenu
+                        bsRole="menu"
+                        className={classNames(className)}
```
don't need call to classNames here since there's only one

alisman commented on src/shared/components/oncoprint/controls/OncoprintControls.tsx:
```diff
@@ -1638,8 +1280,7 @@ export default class OncoprintControls extends React.Component<
                 <ButtonGroup>
-                    <this.ClinicalTracksMenu />
-                    <this.HeatmapMenu />
+                    <this.AddTracksMenu />
```
this is kinda weird.  like, why not just ```{ this.AddTracksMuenu }```.   @adamabeshouse is there a reason?

adamabeshouse replied:
this lets `AddTracksMenu` be an observer component, rather than a computed. I think this gives benefits in compartmentalizing which observables cause changes in various parts of the UI

dippindots commented:
@tmazor I have updated this pr and added the count of the profiles for all tabs (1st suggestion). But the second suggestion we think that would require more discussion. Do you think we can still use the tab name `Heatmap` since we are using `Add Heatmap Tracks` currently? And we can discuss more on this [ticket](https://github.com/cBioPortal/cbioportal/issues/8350). I am sure we will find a good solution to solve all the concerns.

tmazor commented:
@dippindots works for me! the profile counts look good in this PR, and I'm fine with using 'Heatmap' for now and we can use that other ticket to brainstorm other ideas.

adamabeshouse commented on end-to-end-test/shared/specUtils.js:
```diff
@@ -462,6 +462,10 @@ function openGroupComparison(studyViewUrl, chartDataTest, timeout) {
 }
 
+function selectElementByText(text) {
+    return $(`//*[text()="${text}"]`);
```
cool

adamabeshouse commented on src/pages/resultsView/oncoprint/AddTracks.tsx:
```diff
@@ -0,0 +1,476 @@
+    private getTextPixel(text: string, fontSize: string) {
+        // This is a very specified function to calculate the text length in Add Tracks dropdown
+        const FRONT_FAMILY = 'Helvetica Neue';
+        return Math.floor(getTextWidth(text, FRONT_FAMILY, fontSize));
```
👍

adamabeshouse commented:
@dippindots empty file `end-to-end-test/local/screenshots/reference/shows_treatment_profile_heatmap_track_for_treatment_element_chrome_1600x1000.png` is that to be updated from circleCI?

adamabeshouse reviewed (approved):
great!

dippindots commented:
src/pages/resultsView/oncoprint/AddTracksHelper.ts deleted, and all screenshots will display empty message if not select `Display the rich diff`, but those screenshots are updated correctly.
